### ISSUE TITLE
Cannot build chatty

### ISSUE THREAD

**Issue by kuon** [vn971/rua#95]
I tried to install `chatty`, a java app, but the tests fail with the following error `java.lang.NoClassDefFoundError: Could not initialize class sun.awt.X11GraphicsEnvironment`

Is it because the environment inside rua doesn't have access to the X11 session?

**Comment by vn971** [comment]
Yeah, it tries to access X11, which is unavailable. It is a conscious choice to isolate as much as possible when building packages in RUA.

**Comment by vn971** [comment]
Offtopic:
I'm unsure why it's unavailable BTW, because on a default ArchLinux setup, X11 can be accessed via abstract sockets. This is the thing that you can see here: `ss -l | grep @`. It's possible that java doesn't know how to use abstract sockets and only knows how to connect to the normal "file"-like sockets.

There are other apps that can access abstract sockets, and thus your graphical system, even while /tmp is unshared. For those cases, if anybody's interested, you can:
* build in `--offline` mode, which unshares network and disables abstract sockets as a side effect
* start X11 explicitly saying to only accept the normal unix sockets, via `/usr/bin/X -nolisten tcp -nolisten local`, e.g. overriding lightdm settings or whatever other thing is used.

That is, again, an offtopic explaining why some apps can still have access to X11 and how to further prevent that. If the goal is to _have_ X11 access, then it's basically disabled by design. You can temporarily enable it by using `wrap_args` and specifying arguments `--bind /tmp /tmp`

**Comment by vn971** [comment]
Does that answer it / any concerns?

**Comment by kuon** [comment]
I use i3 and start X11 with `startx` so it might be a reason, I still get the socket in `ss-l`

`u_str              LISTEN              0                    128                                                                              @/tmp/.X11-unix/X0 31069  `

I tried using `wrap_args` to mount `/tmp`, it failed at first but then it passed after using `xhost +` to disable access control, I guess that's normal because the cookie is in my home (`~/.Xauthority`).

But I understand about not giving X11 access by default, and agree with this decision (one of the reason I use rua is peace of mind because as a developer I try a gazillion of packages).

I think a line in the README would be a good addition, it wasn't obvious to me at first.

Also, I wonder if it would make sense to find a way to "forward X11" connection, via automatic handling of X11 cookie. To have something like `rua install --forward-x11 <xxx>`. I don't know how common this would be.

**Comment by vn971** [comment]
> I guess that's normal because the cookie is in my home (~/.Xauthority)

It's neater to also mount ~/.Xauthority then: `--ro-bind ~/.Xauthority ~/.Xauthority`. Might also be better as `xhost` is extremely easy to misconfigure in an unsafe way.

`--forward-x11` is entirely possible from a technical perspective. It would just automatically bind: `--bind /tmp/.X11-unix /tmp/.X11-unix --ro-bind ~/.Xauthority ~/.Xauthority`. Not sure how common that'll be...:thinking:

**Comment by vn971** [comment]
Do you remember if there's maybe some "virtual" X11 server? Like Xephyr that nests apps into a parent X11, but then a fully virtual one that doesn't nest anything anywhere at all. But still usable for building/testing X11 apps.

**Comment by kuon** [comment]
Yeah of course, `xhost +` was just to make it work, mounting `.Xauthority` is the way to go.

As for dummy X11, there is `xf86-video-dummy` and `xf86-video-nested` which is the modern version of `Xvfb` and `Xzephyr`.

Using X11 dummy is done like this:

`dummy.conf`

```
    Section "Monitor"
      Identifier "Monitor0"
      HorizSync 28.0-80.0
      VertRefresh 48.0-75.0
      # https://arachnoid.com/modelines/
      # 1920x1080 @ 60.00 Hz (GTF) hsync: 67.08 kHz; pclk: 172.80 MHz
      Modeline "1920x1080_60.00" 172.80 1920 2040 2248 2576 1080 1081 1084 1118 -HSync +Vsync
    EndSection
    Section "Device"
      Identifier "Card0"
      Driver "dummy"
      VideoRam 256000
    EndSection
    Section "Screen"
      DefaultDepth 24
      Identifier "Screen0"
      Device "Card0"
      Monitor "Monitor0"
      SubSection "Display"
        Depth 24
        Modes "1920x1080_60.00"
      EndSubSection
    EndSection

```

Then start `X11` with `/usr/lib/Xorg :1 -config dummy.conf`.

**Comment by vn971** [comment]
The tool `Xvfb` is mentioned as "deprecated" on wikipedia, but it seems to work quite well with providing a virtual, separated X11. You can start it just from the command line via `Xvfb :2`.

We can then maybe introduce a flag like `--virtual-x11` to automatically create a virtual X11 (whatever that'll be), with a predefined DISPLAY number, and pass that environment variable to descending processes. What do you think of the idea?

**Comment by vn971** [comment]
Note: you'll probably have to work on such a feature on your own, should you wish to. At least until I or many other people stumble across the same thing.

**Comment by kuon** [comment]
I suggest we keep this issue open for now, and if other people mention interest in it I'll try to implement it.

I think it could be of interest for CI mostly.

**Comment by vn971** [comment]
I've documented X11 access restriction [here](https://github.com/vn971/rua/blob/master/docs/x11access.md).
